Plugins - High level SDK features (#530)
strands-agents/docs: designs/0007-plugins.md

## Adding a new concept to Strands - Plugins

Strands SDK has defined its **low-level primitives** in such a way that any of these "agent-building techniques" are possible to implement, but there are issues with the recommended path.

The current recommended approach in the sdk to add new functionality is to either introduce a new initialization parameter to the agent, or create a `HookProvider` which introduces this functionality. We cannot reasonably add `n` number of new parameters to the Agent, so the best path forward is to start creating and vending new `HookProvider` in the sdk. 

Hook Providers ultimately solve our problem since they are composable, provide the necessary functionality to integrate rich features throughout an agents execution, and have been battle tested by many of our existing **high-level abstractions**.

> One problem with `HookProviders` is that their name does not best communicate what they are positioned to do. Their name overlaps with the **low-level primitives** hooks, so it can be not-so intuitive when applying a high-level feature, like Anthropic Skills, through a `hooks` parameter (see code below). I propose that we create a new `Plugin` concept, and a `plugins` agent init parameter, as plugins are a common term in the industry to represent adding new functionality to a system.

Reviewer: nit: skills is also cross-component. it'd be hooks, system prompt, tools, etc.

That being said, some features have a large impact on a majority of agents, so giving them a top-level agent initialization parameter help distinguish them from the feature-set of other `Plugins`. This also helps signal that they are "mostly" cross-compatible with any other `Plugin`.

A good example of this is the Session Manager, where persisting and restoring an agent is a common and useful feature, and can contain additional functionality of persisting the state of other plugins ([see Appendix](#persisting-the-state-of-plugins)). We can decide to uplevel a plugin to an Agent class attribute based on its use case.

### Hooks vs HookProvider vs Plugins
These features all have similar responsibilities, and the differences between them are nuanced, so here are definitions to explain what they do:

- **Hooks**: A low-level primitive of the Strands SDK. A hook is a mechanism to execute code at a specific lifecycle event in the SDK, and gives relevant context at that specific lifecycle event in order to change the standard agents behavior.

> - **HookProvider**: A provider of hooks, with some convenience methods to make it easy to apply hooks to an agent.

Reviewer: I see HookProvider as just an implementation detail. With the `@hook` decorator, that should just be sth behind the scenes

See [Proposed SDK Architecture Changes](#proposed-sdk-architecture-changes) to better understand how the code will change.

## Proposed SDK Architecture Changes

Add a `plugins` agent init parameter, and deprecate `hooks`.

```python

class Plugin(Protocol):
    def init_plugin(self, agent: Agent):
        ...

agent = Agent(
    plugins=[
        SkillsPlugin(),
        CodeActPlugin()
    ],
    hooks=[]    # Deprecated
)

# Simple hook can still be defined as a plugin with a decorator
@hook
def log_calls(event: BeforeModelCallEvent): 
    print(f"Calling model...")

# Usage
agent = Agent(
    plugins=[log_calls]
)

# Or applied with direct-access
agent.hooks.add_callback(BeforeInvocationEvent, log_calls)
```

With this approach, we deprecate `HookProviders` and the top-level `hooks` parameter, introduce a new `Plugin` protocol and `plugins` agent init arg, and start vending Agent behavior changing features as `Plugins` (like steering).

An easy way to introduce low-level hooks is still valuable, so we can introduce a `@hook` decorator where you can wrap a function as a simple plugin.

# Appendix

## Persisting the state of Plugins

> The Session Manager has the responsibility of persisting and restoring the state of an agent, along with any attached plugins. The recommended way to accomplish this for now would be to have plugins maintain their state in the `agent.state`. Then during initialization, session manager would restore `agent.state`, and each plugin could restore themselves in turn from `agent.state`.

Reviewer: can't we just have those support snapshottable interface? if they want to? we can just load them back

# does a better job communicating the feature
agent = Agent(hooks=[AnthropicSkills()])

agent = Agent(plugins=[AnthropicSkills()])
```
**Note**: `Extension` is another option, but I am just biased toward plugins. Either way, we need a new named term to represent the introduction of these features in the sdk.

I go into more detail about [Plugins and Hooks](#plugins-and-hooks) below.

### What is a `Plugin`?

`Plugins` provide a mechanism for customers to apply a behavior change (like Anthropic Skills) to an agent in a way that is extensible, composable, and shareable.

#### Extensible
We will provide an abstraction for customers to implement their own plugins. While we may ship some with the sdk, customers are welcome to develop their own as well.

#### Composable
Multiple `Plugins` can be applied to a single agent. While we cannot guarantee that these plugins will not conflict with one another, it leads to maximum flexibility over what they can do.

#### Sharable
Similar to our guidance on **low-level primitives**, we will recommend that community members share their own plugins by vending their own packages through some distributions mechanism (pypi, npm, github, etc).

> ### What isn't a `Plugin`?

Reviewer 2: Got it. How do we handle it if a plugin overwrites the session manager the customer used, for example? Does this mean plugins will need to have cross-compatibility or certify they work with other things?

> ### What isn't a `Plugin`?

Author: We can incrementally work toward this, but as a P0 I dont think we should build those guarantees in. I mention this in the Composable section:
```
Multiple `Plugins` can be applied to a single agent. While we cannot guarantee that these plugins will not conflict with one another, it leads to maximum flexibility over what they can do.
```

We can probably do better, but before implementing them, I think its fine to lean into this flexibility

Today Strands ships with many **high-level abstractions** that the team has determined to be valuable to a majority of customers:
- Session Manager
- OTEL
- Conversation Manager
- RetryStrategy
- Steering

All of these features have been built into the sdk before the introduction of the concept of `Plugins`, but if they were built today, would potentially have been implemented as a `Plugin`. So this begs the question - What is or isn't a `Plugin`?

From the backwards compatibility perspective, we cannot just remove these features and make turn them into plugins, but we can re-think how we categorize these features. If a feature needs to perform an action that cannot be accomplished with the existing **low-level primitives** offered by the sdk, then the feature would likely not be considered a plugin. OTEL, for example, is a feature that may need to operate behind the scenes of the existing primitives (creating traces around the execution of primitives), so it does not fit well into the `Plugin` terminology. However the rest of the features mentioned in the list above can be reasonably defined as a `Plugin`.

> That being said, some features have a large impact on a majority of agents, so giving them a top-level agent initialization parameter help distinguish them from the feature-set of other `Plugins`. This also helps signal that they are "mostly" cross-compatible with any other `Plugin`.

Reviewer 3: Two additional deciding factors:

 - When a plugin is meant to be exclusive - e.g. you can only have "one" - two session-managers doesn't make sense because they clobber eachother
 - When a plugin modifies existing behavior - e.g. the retry-strategy.  In that case a plugin doesn't make sense because how do you indicate that you're adjusting values vs being "additive"

Author: Agree with the first point. But for retry strategy, Im a bit more on the fence. I could see someone adding an `AnthropicRetryStrategy` and an `OpenAIRetryStrategy` to the same agent through a plugin.

    def log_end(self, event: AfterInvocationEvent) -> None:
        print(f"Request completed for agent: {event.agent.name}")

# Passed in via the hooks parameter
agent = Agent(plugins=[LoggingPlugin()])
```

The usage of each abstraction is essentially the same, so in order to avoid the customer confusion of which to choose, I'm proposing that we deprecate `HookProvider` in favor of `Plugin` moving forward. There is a world where `Plugins` can depend on `HookProviders`, and extend the behavior, but I think this leads to more indirection rather than proper abstractions. 

That being said, we are not deprecating the concepts of `hooks` in the SDK, which is the **low-level primitive** for running code at lifecycle events of an agent. A mechanism to easily write targeted `hooks` should also be included so developers can write a hook in a few lines of code, like the `@hooks` decorator [proposed in this pr](https://github.com/strands-agents/sdk-python/pull/1581):

```python
# Simple hook can still be defined as a plugin with a decorator
@hook
def log_calls(event: BeforeModelCallEvent): 
    print(f"Got event: {event}")

# Usage
agent = Agent(

>     plugins=[log_calls]

Reviewer 4: Exactly, if @hook is just creating a plugin, it should be @plugin. But, we know that the concept of a hook still exists so we need to figure out how that fits in cleanly.

>     plugins=[log_calls]

Author: ```
from strands import hook

class LoggerPlugin(Plugin):

  @hook
  def logger(event: BeforeModelCallEvent)
  ..

@hook
def logger(event: BeforeModelCallEvent | AfterModelCallEvent):
  print(event)
```

This is our current best idea, but we are worried that the `@hook` decorator does not exactly imply that this is a plugin.

Anthropic skills represent one of the many ways to build an Agent, but with some other popular examples being: Steering, Code-Act, [Ralph Wiggum](https://awesomeclaude.ai/ralph-wiggum), Memory (Agentcore or otherwise), and many many more. Additionally, these concepts don't need to exist in isolation; You can build an agent that uses both Anthropic Skills and Steering.

## Adding a new concept to Strands - Plugins

Strands SDK has defined its **low-level primitives** in such a way that any of these "agent-building techniques" are possible to implement, but there are issues with the recommended path.

The current recommended approach in the sdk to add new functionality is to either introduce a new initialization parameter to the agent, or create a `HookProvider` which introduces this functionality. We cannot reasonably add `n` number of new parameters to the Agent, so the best path forward is to start creating and vending new `HookProvider` in the sdk. 

> Hook Providers ultimately solve our problem since they are composable, provide the necessary functionality to integrate rich features throughout an agents execution, and have been battle tested by many of our existing **high-level abstractions**.

Author: These are mentioned earlier, but steering or session manager are good examples.

One problem with `HookProviders` is that their name does not best communicate what they are positioned to do. Their name overlaps with the **low-level primitives** hooks, so it can be not-so intuitive when applying a high-level feature, like Anthropic Skills, through a `hooks` parameter (see code below). I propose that we create a new `Plugin` concept, and a `plugins` agent init parameter, as plugins are a common term in the industry to represent adding new functionality to a system.

```python
# Between these two examples, I think the plugins example
# does a better job communicating the feature
agent = Agent(hooks=[AnthropicSkills()])

agent = Agent(plugins=[AnthropicSkills()])
```

> **Note**: `Extension` is another option, but I am just biased toward plugins. Either way, we need a new named term to represent the introduction of these features in the sdk.

Reviewer 3: Worth double-checking with product about naming in case they have a preference.

I go into more detail about [Plugins and Hooks](#plugins-and-hooks) below.

### What is a `Plugin`?

`Plugins` provide a mechanism for customers to apply a behavior change (like Anthropic Skills) to an agent in a way that is extensible, composable, and shareable.

#### Extensible
We will provide an abstraction for customers to implement their own plugins. While we may ship some with the sdk, customers are welcome to develop their own as well.

#### Composable

> Multiple `Plugins` can be applied to a single agent. While we cannot guarantee that these plugins will not conflict with one another, it leads to maximum flexibility over what they can do.

Reviewer 3: ```suggestion
Multiple `Plugins` can be applied to a single agent. We make no garantees around plugins not conflicting as to provide maximum flexibility over what they can do.
```

I'd almost say preventing incompatibility  an anti-goal, but a good goal would be allowing plugins the ability to query what other plugins are available.  This is used in Astro/Starlight for not adding functionality when a known plugin is installed, for example

Reviewer 3: I think this also goes into "what's a plugin" vs "what's a parameter" later - e.g. when we identify an area as important enough or is something where only a single "Component" should adjust, that's when it's no longer a plugin but a higher level concept

#### Sharable
Similar to our guidance on **low-level primitives**, we will recommend that community members share their own plugins by vending their own packages through some distributions mechanism (pypi, npm, github, etc).

### What isn't a `Plugin`?

Today Strands ships with many **high-level abstractions** that the team has determined to be valuable to a majority of customers:
- Session Manager
- OTEL
- Conversation Manager
- RetryStrategy
- Steering

All of these features have been built into the sdk before the introduction of the concept of `Plugins`, but if they were built today, would potentially have been implemented as a `Plugin`. So this begs the question - What is or isn't a `Plugin`?

> From the backwards compatibility perspective, we cannot just remove these features and make turn them into plugins, but we can re-think how we categorize these features. If a feature needs to perform an action that cannot be accomplished with the existing **low-level primitives** offered by the sdk, then the feature would likely not be considered a plugin. OTEL, for example, is a feature that may need to operate behind the scenes of the existing primitives (creating traces around the execution of primitives), so it does not fit well into the `Plugin` terminology. However the rest of the features mentioned in the list above can be reasonably defined as a `Plugin`.

Reviewer 5: is renaming still not backwards incompatible?

That being said, some features have a large impact on a majority of agents, or only make sense if there is "one" instance of it, so giving them a top-level agent initialization parameter help distinguish them from the feature-set of other `Plugins`. This also helps signal that they are "mostly" cross-compatible with any other `Plugin`.

A good example of this is the Session Manager, where persisting and restoring an agent is a common and useful feature, and can contain additional functionality of persisting the state of other plugins ([see Appendix](#persisting-the-state-of-plugins)). We can decide to uplevel a plugin to an Agent class attribute based on its use case.

## Proposed SDK Changes

Add a `plugins` agent init parameter, and deprecate `hooks`.

```python

> class Plugin(Protocol):

Reviewer: is this the final protocol?

Author: This is the rough shape. I dont want to do implementation review here, this is just meant to get agreement on roughly what a plugin does.

Reviewer 6: let's dig in to that when appropriate. The current interface is very thin which raises some questions for me

    name: str

>     def init_plugin(self, agent: Agent):

Reviewer 3: Probably makes sense to be async - specifically for session-manager purposes? (I'm thinking TS for example)

>     def init_plugin(self, agent: Agent):

Author: We talked about having the ability to add hooks easily to the plugin:
```
class LoggerPlugin(Plugin):
  @hook
  def log(event: BeforeModelCallEvent):
    print(event)
```

But agree, we will explore this more during implementation.

        SkillsPlugin(),
        CodeActPlugin()
    ],
    hooks=[]    # Deprecated
)

```

With this approach: 
- We deprecate `HookProviders` (see [Why not HookProviders? below](#why-not-hookproviders)) and the top-level `hooks` parameter
- Introduce a new `Plugin` protocol and `plugins` agent initialization argument
- Start vending Agent behavior changing features as `Plugins` (like steering, skills, context management, etc).

### Additional changes

#### Introduce `add_hook` method to the agent class

An `add_hook` method is easier to write, and is just as clear as the previous method for registers a callback to the specified lifecycle event.
```python
# Before
agent.hooks.add_callback(BeforeInvocationEvent, self.log_start)
# After
agent.add_hook(BeforeModelCallEvent, callback)
```

#### Simple `hook` callback authoring
A `hook` registered on an agent will invoke a callback function at a specified lifecycle event. While a user can simply define an agent, then use the `agent.add_hook` method to attach the hook to the agent, there isn't a simple way to create attach a `hook` during agent initialization. Lets looks at a simple logging use case:

```python
# Plugin example

> class LoggingPlugin(Plugin):

Reviewer: are hooks plugins? 

can I do

```python

@hook
def myhook(event: ...):
   ...

agent(plugins=[myhook])
```

Reviewer: nvm saw it later :p

    # Protocol method for a Plugin
    def init_plugin(self, agent: Agent) -> None:
        agent.add_hook(BeforeInvocationEvent, self.log_start)

    def log_start(self, event: BeforeInvocationEvent) -> None:
        print(f"Request started for agent: {event.agent.name}")

agent = Agent(plugins=[LoggingPlugin()])

# Post agent initialization example
def log_start(self, event: BeforeInvocationEvent) -> None:
    print(f"Request started for agent: {event.agent.name}")

agent = Agent()
agent.add_hook(BeforeInvocationEvent, log_start)
```

While both of these examples attach a hook to the agent, there is room for improvement to meet our [We are accessible to humans and agents](https://github.com/strands-agents/sdk-python/blob/main/CONTRIBUTING.md?plain=1#L46) tenet where we say, "We don’t take shortcuts on curated DX for humans".

> To help improve this devex, a new `@hook_callback` decorator can be introduced to make it easy to wrap a function as a plugin, and then pass it into agent initialization. This aligns more closely with our "curated DX for humans" that the Strands team attempts to excel in.

Reviewer 3: meh - not sure it really matters that much

Maybe my larger point is - do we need a streamed line way of creating a plugin instead?  Like a function that is called for all hooks? (callback_handler).

> To help improve this devex, a new `@hook_callback` decorator can be introduced to make it easy to wrap a function as a plugin, and then pass it into agent initialization. This aligns more closely with our "curated DX for humans" that the Strands team attempts to excel in.

Author: Renaming this to @hook for now, we will continue to explore what the best devex is here.

```python

> @hook_callback

Author: Going with @hook for now, but its not decided on yet. Will iterate before including.

def log_calls(event: BeforeModelCallEvent): 
    print(f"Calling model...")

> agent = Agent(plugins=[log_calls])

Reviewer 7: So is this saying that a plugin could be a Plugin instance or a hook callback?

    def log_end(self, event: AfterInvocationEvent) -> None:
        print(f"Request completed for agent: {event.agent.name}")

agent = Agent(hooks=[LoggingHook()])

# Plugin example
# Note: Shape here is an example, not finalized
class LoggingPlugin(Plugin):
    # Protocol method for a Plugin
    def init_plugin(self, agent: Agent) -> None:
        agent.add_hook(BeforeInvocationEvent, self.log_start)
        agent.add_hook(AfterInvocationEvent, self.log_end)

    def log_start(self, event: BeforeInvocationEvent) -> None:
        print(f"Request started for agent: {event.agent.name}")

    def log_end(self, event: AfterInvocationEvent) -> None:
        print(f"Request completed for agent: {event.agent.name}")

# Passed in via the hooks parameter
agent = Agent(plugins=[LoggingPlugin()])
```

The usage of each abstraction is essentially the same, so in order to avoid the customer confusion of which to choose, I'm proposing that we deprecate `HookProvider` in favor of `Plugin` moving forward. There is a world where `Plugins` can depend on `HookProviders`, and extend the behavior, but I think this leads to more indirection rather than proper abstractions. 

> To be clear, we are not deprecating the concepts of `hooks` in the SDK, which is the **low-level primitive** for running code at lifecycle events of an agent. We are just renaming the previous abstraction of `HookProvider` to `Plugin` to better communicate its intended functionality to customers.

Reviewer 8: Maybe move it to the top? it is a foundimental idea
